README
IO-All-LWP version 0.13
=======================

This module acts as glue between IO::All and LWP, so that files can be read and
written through the network using the convenient IO:All interface.  Note that
this module is not "use"d directly: you just use IO::All, which knows when to
autoload IO::All::HTTP, IO::All::HTTPS, IO::All::FTP, or IO::All::Gopher, which
implement the specific protocols based on IO::All::LWP.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:
    perl 5.8.0
    IO::All 0.30
    LWP 0

COPYRIGHT AND LICENSE

Copyright (C) 2004 Ivan Tubert-Brohman <itub@cpan.org> and Brian Ingerson
<ingy@cpan.org>

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself. 



Changes
Revision history for IO-All-LWP

0.13  2007-03-28
        - Updated for new version of IO::All. Thanks to MAREKR for sending a
          patch!

0.12  2004-11-02
        - Added IO::All::LWP module. Thanks Richard Soderberg!

0.11  2004-10-1
        - Added ua method.
        - Fixed error in the docs.

0.10  2004-07-20
	- First release


lib/IO/All/FTP.pm
package IO::All::FTP;
use strict;
use warnings;
our $VERSION = '0.13';
use IO::All::LWP '-base';

const type => 'ftp';

sub ftp { my $self=shift; $self->lwp_init(__PACKAGE__, @_) }

1;

__END__

=head1 NAME

IO::All::FTP - Extends IO::All to FTP URLs

=head1 SYNOPSIS

    use IO::All;

    "hello world\n" > io('ftp://localhost/test/x');  # save to FTP
    io('ftp//example.org/pub/xyz') > io('xyz');      # GET to file

    # two ways of getting a file with a password:
    $content < io('ftp://me:secret@example.org/xyz');
    $content < io('ftp://example.org/xyz')->user('me')->password('secret');

=head1 DESCRIPTION

This module extends IO::All for dealing with FTP URLs.
Note that you don't need to use it explicitly, as it is autoloaded by
L<IO::All> whenever it sees something that looks like an FTP URL.

=head1 METHODS

This is a subclass of L<IO::All::LWP>. The only new method is C<ftp>, which
can be used to create a blank L<IO::All::FTP> object; or it can also take an
FTP URL as a parameter. Note that in most cases it is simpler just to call
io('ftp//example.com'), which calls the C<ftp> method automatically.

=head1 OPERATOR OVERLOADING

The same operators from IO::All may be used. < GETs an FTP URL; > PUTs to
an FTP URL.

=head1 SEE ALSO

L<IO::All::LWP>, L<IO::All>, L<LWP>.

=head1 AUTHORS

Ivan Tubert-Brohman <itub@cpan.org> and 
Brian Ingerson <ingy@cpan.org>

=head1 COPYRIGHT

Copyright (c) 2007. Ivan Tubert-Brohman and Brian Ingerson. All rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

See L<http://www.perl.com/perl/misc/Artistic.html>

=cut



lib/IO/All/HTTP.pm
package IO::All::HTTP;
use strict;
use warnings;
our $VERSION = '0.13';
use IO::All::LWP '-base';
use URI::http; 
use URI::_userpass; 
push @URI::http::ISA, 'URI::_userpass';

const type => 'http';

sub http { my $self=shift; $self->lwp_init(__PACKAGE__, @_) }

1;

__END__

=head1 NAME

IO::All::LWP - Extends IO::All to HTTP URLs

=head1 SYNOPSIS

    use IO::All;

    $content < io('http://example.org');             # GET webpage into scalar
    io('http://example.org') > io('index.html');     # GET to file
    "hello\n" > io('http://example.org/index.html'); # PUT webpage

    # two ways of getting a page with a password:
    $content < io('http://me:secret@example.org');
    $content < io('http://example.org')->user('me')->password('secret');

=head1 DESCRIPTION

This module extends L<IO::All> for dealing with HTTP URLs. 
Note that you don't need to use it explicitly, as it is autoloaded by
L<IO::All> whenever it sees something that looks like an HTTP URL.

The SYNOPSIS shows some simple typical examples, but there are many other
interesting combinations with other IO::All features! For example, you can get
an HTTP URL and write the content to a socket, or to an FTP URL, of to a DBM
file.

=head1 METHODS

This is a subclass of L<IO::All::LWP>. The only new method is C<http>, which
can be used to create a blank L<IO::All::HTTP> object; or it can also take an
HTTP URL as a parameter. Note that in most cases it is simpler just to call
io('http://example.com'), which calls the C<http> method automatically.

=head1 OPERATOR OVERLOADING

The same operators from IO::All may be used. < GETs an HTTP URL; > PUTs to
an HTTP URL.

=head1 SEE ALSO

L<IO::All>, L<IO::All::LWP>, L<LWP>.

=head1 AUTHORS

Ivan Tubert-Brohman <itub@cpan.org> and 
Brian Ingerson <ingy@cpan.org>

=head1 COPYRIGHT

Copyright (c) 2007. Ivan Tubert-Brohman and Brian Ingerson. All rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

See L<http://www.perl.com/perl/misc/Artistic.html>

=cut



lib/IO/All/HTTPS.pm
package IO::All::HTTPS;
use strict;
use warnings;
our $VERSION = '0.13';
use IO::All::HTTP '-base';

const type => 'https';

sub https { my $self=shift;$self->lwp_init(__PACKAGE__, @_) }

1;

__END__

=head1 NAME

IO::All::HTTPS - Extends IO::All for HTTPS URLs

=head1 SYNOPSIS

    use IO::All;

    $content < io('https://example.org');   # GET webpage

    # two ways of getting a page with a password:
    $content < io('https://me:secret@example.org');
    $content < io('https://example.org')->user('me')->password('secret');


=head1 DESCRIPTION

This module extends L<IO::All> for dealing with HTTPS URLs. 
Note that you don't need to use it explicitly, as it is autoloaded by
L<IO::All> whenever it sees something that looks like an HTTPS URL.

The SYNOPSIS shows some simple typical examples, but there are many other
interesting combinations with other IO::All features! For example, you can get
an HTTPS URL and write the content to a socket, or to an FTP URL, of to a DBM
file.

=head1 METHODS

This is a subclass of L<IO::All::LWP>. The only new method is C<https>, which
can be used to create a blank L<IO::All::HTTPS> object; or it can also take an
HTTPS URL as a parameter. Note that in most cases it is simpler just to call
io('https://example.com'), which calls the C<https> method automatically.

=head1 OPERATOR OVERLOADING

The same operators from IO::All may be used. < GETs an HTTPS URL; > PUTs to
an HTTPS URL.

=head1 SEE ALSO

L<IO::All>, L<IO::All::LWP>, L<LWP>.

=head1 AUTHORS

Ivan Tubert-Brohman <itub@cpan.org> and 
Brian Ingerson <ingy@cpan.org>

=head1 COPYRIGHT

Copyright (c) 2007. Ivan Tubert-Brohman and Brian Ingerson. All rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

See L<http://www.perl.com/perl/misc/Artistic.html>

=cut



lib/IO/All/LWP.pm
package IO::All::LWP;
require 5.008;
use strict;
use warnings;
our $VERSION = '0.13';
use IO::All 0.30 '-base';
use LWP::UserAgent;
use IO::Handle;

my $DEFAULT_UA = LWP::UserAgent->new(env_proxy => 1);

field 'response';
field 'content';
field 'put_content';

sub lwp_init {
    bless my $self, shift;
    $self->name(shift) if @_;
    return $self->_init;
}

sub ua {
    my $self = shift;
    if (@_) {
        *$self->{ua} = ref($_[0]) ? shift :
            LWP::UserAgent->new(@_);
        return $self;
    } else {
        *$self->{ua} ||= $DEFAULT_UA;
    }
}

sub uri {
    my $self = shift;
    *$self->{uri} = ref($_[0]) ? shift : URI->new(shift)
      if @_;
    return *$self->{uri}
      if defined *$self->{uri};
    *$self->{uri} = URI->new($self->name);
}
    
sub user {
    my $self = shift;
    $self->uri->user(@_);
    return $self;
}

sub password {
    my $self = shift;
    $self->uri->password(@_);
    return $self;
}

sub get {
    my $self = shift;
    my $request = shift || HTTP::Request->new('GET', $self->uri);
    $self->request($request);
}

sub put {
    my $self = shift;
    my $request = (@_ and ref $_[0])
    ? $_[0]
    : do {
        my $content = @_ ? shift : $self->content;
        HTTP::Request->new(PUT => $self->uri, undef, $content);
    };
    $self->request($request);
    $self->is_open(0);
}

sub request {
    my $self = shift;
    $self->response($self->ua->request(shift));
}

sub open {
    my $self = shift;
    $self->is_open(1);
    my $mode = @_ ? shift : $self->mode ? $self->mode : '<';
    $self->mode($mode);
    my $fh;
    if ($mode eq '<') {
        $self->content($self->get->content);
        CORE::open $fh, "<", \ $self->content;
    } 
    elsif ($mode eq '>') {
        $self->put_content(\ do{ my $x = ''});
        CORE::open $fh, ">", $self->put_content;
    } 
    $self->io_handle($fh);
    return $self;
}

sub close {
    my $self = shift;
    if ($self->is_open and defined $self->mode and $self->mode eq '>') {
        $self->content(${$self->put_content});
        $self->put;
    }
    $self->SUPER::close;
}

1;
__END__

=head1 NAME

IO::All::LWP - IO::All interface to LWP

=head1 SYNOPSIS

    use IO::All;

    "hello world\n" > io('ftp://localhost/test/x');   # save to FTP
    $content < io('http://example.org');              # GET webpage

    io('http://example.org') > io('index.html');      # save webpage

=head1 DESCRIPTION

This module acts as glue between L<IO::All> and L<LWP>, so that files can be
read and written through the network using the convenient L<IO:All> interface.
Note that this module is not C<use>d directly: you just use L<IO::All>, which
knows when to autoload L<IO::All::HTTP>, L<IO::All::HTTPS>, L<IO::All::FTP>, or
L<IO::All::Gopher>, which implement the specific protocols based on
L<IO::All::LWP>.

=head1 EXECUTION MODEL

B<GET requests>. When the IO::All object is opened, the URI is fetched and
stored by the object in an internal file handle. It can then be accessed like
any other file via the IO::All methods and operators, it can be tied, etc.

B<PUT requests>. When the IO::All object is opened, an internal file handle is
created. It is possible to that file handle using the various IO::All methods
and operators, it can be tied, etc. If $io->put is not called explicitly, when
the IO::All object is closed, either explicitly via $io->close or automatically
upon destruction, the actual PUT request is made.

The bad news is that the whole file is stored in memory after getting it or
before putting it. This may cause problems if you are dealing with
multi-gigabyte files!

=head1 METHODS

The simplest way of doing things is via the overloaded operators > and <, as
shown in the SYNOPSIS. These take care of automatically opening and closing the
files and connections as needed. However, various methods are available to
provide a finer degree of control. 

This is a subclass of L<IO::All>. In addition to the inherited methods, the 
following methods are available:

=over

=item * ua

Set or get the user agent object (L<LWP::UserAgent> or a subclass). If called
with a list, the list is passed to LWP::UserAgent->new. If called with an
object, the object is used directly as the user agent. Note that there is a 
default user agent if no user agent is specified.

=item * uri

Set or get the URI. It can take either a L<URI> object or a string, and 
it returns an L<URI> object. Note that calling this method overrides the user
and password fields, because URIs can contain authentication information.

=item * user

Set or get the user name for authentication. Note that the user name (and the
password) can also be set as part of the URL, as in
"http://me:secret@example.com/".

=item * password

Set or get the password for authentication. Note that the password can also
be set as part of the URL, as discussed above.

=item * get

GET the current URI using LWP. Or, if called with an L<HTTP::Request> object as
a parameter, it does that request instead. It returns the L<HTTP::Response>
object.

=item * put

PUT to the current URI using LWP. If called with an L<HTTP::Request> object, it
does that request instead. If called with a scalar, it PUTs that as the
content to the current URI, instead of the current accumulated content.

=item * response

Return the L<HTTP::Response> object.

=item * request

Does an LWP request. It requires an L<HTTP::Request> object as a parameter.
Returns an L<HTTP::Response> object.

=item * open

Overrides the C<open> method from L<IO::All>. It takes care of GETting the 
content, or of setting up the internal buffer for PUTting. Just like the
C<open> method from L<IO::All>, it can take a mode: '<' for GET and 
'>' for PUT.

=item * close

Overrides the C<close> method from L<IO::All>. It takes care of PUTting
the content.

=back

=head1 DEPENDENCIES

This module uses L<LWP> for all the heavy lifting. It also requires perl-5.8.0
or a more recent version.

=head1 SEE ALSO

L<IO::All>, L<LWP>, L<IO::All::HTTP>, L<IO::All::FTP>.

=head1 AUTHORS

Ivan Tubert-Brohman <itub@cpan.org> and 
Brian Ingerson <ingy@cpan.org>

Thanks to Sergey Gleizer for the ua method.

=head1 COPYRIGHT

Copyright (c) 2007. Ivan Tubert-Brohman and Brian Ingerson. All rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

See L<http://www.perl.com/perl/misc/Artistic.html>

=cut

